File: app/src/main/java/com/cbr/behancesampleapp/ui/components/PagingAdapter.java
package com.cbr.behancesampleapp.ui.components;

import android.support.v7.widget.RecyclerView;
import android.util.Log;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Created by Dimitrios on 8/26/2017.
 */

public abstract class PagingAdapter<T, V extends RecyclerView.ViewHolder> extends RecyclerView.Adapter<V> {

	private ArrayList<PagingListItem<T>> mData = new ArrayList<>();

	private final static int PAGING_THREESHOLD = 7;

	private Interactor mInteractor;
	private PagingListItem<T> mLoadingSpinner = new PagingListItem<>(null, PagingListItem.Type.LOADING);

	private boolean mIsLoading;
	private boolean mIsFinished;

	public void setInteractor(Interactor interactor) {
		this.mInteractor = interactor;
	}

	@Override
	public int getItemCount() {
		return mData.size() + 1;
	}

	@Override
	public int getItemViewType(int position) {
		return position == mData.size() ? PagingListItem.Type.LOADING.ordinal() : mData.get(position).getType().ordinal();
	}

	public ArrayList<PagingListItem<T>> getData() {
		return this.mData;
	}

	@Override
	public void onBindViewHolder(V holder, int position, List<Object> payloads) {
		super.onBindViewHolder(holder, position, payloads);
		if (mInteractor != null && position > mData.size() - PAGING_THREESHOLD && !(mIsLoading || mIsFinished)) {
			startLoading();
			mInteractor.requestMoreData();
		}
	}

	public void onDataLoaded(Collection<T> data, boolean clearPrevious) {
		stopLoading();
		if (clearPrevious) {
			clearData();
		}
		addData(data);
	}

	private void addData(Collection<T> data) {
		if (data != null && !data.isEmpty()) {
			int currentItems = mData.size();
			int newItemsCounter = 0;
			for (T item : data) {
				getData().add(currentItems + newItemsCounter, new PagingListItem<T>(item, PagingListItem.Type.DATA));
				newItemsCounter++;
			}
			notifyItemRemoved(currentItems);
			notifyItemRangeInserted(currentItems, data.size());
		}
	}

	public void clearData() {
		int previousCount = mData.size();
		getData().clear();
		if (previousCount > 0) {
			notifyItemRangeRemoved(0, previousCount);
		}
	}

	public void startLoading() {
		mIsLoading = true;
	}

	public void stopLoading() {
		mIsLoading = false;
	}

	public void onAllItemsLoaded() {
		mIsFinished = true;
	}

	public interface Interactor {

		void requestMoreData();

	}

}

File: app/src/main/java/com/cbr/behancesampleapp/ui/common/recycler/PagingAdapter.kt
package com.cbr.behancesampleapp.ui.common.recycler

import android.support.v7.widget.RecyclerView

/** Created by Dimitrios on 8/26/2017. */
abstract class PagingAdapter<T, V : RecyclerView.ViewHolder>(val interactor: Interactor<T>) : RecyclerView.Adapter<V>() {
    
    val PAGING_THREESHOLD = 7
    
    val data = mutableListOf<PagingListItem<T>>()
    
    private var mIsLoading: Boolean = false
    private var mIsFinished: Boolean = false
    
    override fun getItemCount(): Int = data.size + 1
    
    override fun getItemViewType(position: Int): Int =
            if (position == data.size) PagingListItem.Type.LOADING.ordinal
            else data[position].type.ordinal
    
    override fun onBindViewHolder(holder: V, position: Int, payloads: MutableList<Any>) {
        super.onBindViewHolder(holder, position, payloads)
        if (position > data.size - PAGING_THREESHOLD && !(mIsLoading || mIsFinished)) {
            startLoading()
            interactor.requestMoreData()
        }
    }
    
    fun onDataLoaded(data: Collection<T>, clearPrevious: Boolean) {
        stopLoading()
        if (clearPrevious) {
            clearData()
        }
        addData(data)
    }
    
    private fun addData(items: Collection<T>?) {
        if (items != null && !items.isEmpty()) {
            val currentItems = this.data.size
            var newItemsCounter = 0
            for (item in items) {
                data.add(currentItems + newItemsCounter, PagingListItem(item, PagingListItem.Type.DATA))
                newItemsCounter++
            }
            notifyItemRemoved(currentItems)
            notifyItemRangeInserted(currentItems, data.size)
        }
    }
    
    fun clearData() {
        val previousCount = data.size
        data.clear()
        if (previousCount > 0) {
            notifyItemRangeRemoved(0, previousCount)
        }
    }
    
    fun startLoading() {
        mIsLoading = true
    }
    
    fun stopLoading() {
        mIsLoading = false
    }
    
    fun onAllItemsLoaded() {
        mIsFinished = true
    }
    
    interface Interactor<T> {
        
        fun requestMoreData()
        
        fun onListItemClicked(item: T, position: Int)
        
    }
}
